In [71]:
import pandas as pd
import numpy as np
import datetime as dt

In [72]:
# Произведем загрузку датасетов и анализ типов содержащихся данных
founders = pd.read_csv('founders.csv')

In [73]:
#создание функции выводящей все характеристики таблицы на экран
def first_check(dataset):

    print()
    print('Первые 3 строки таблицы')
    display(dataset.head(3))

    print()
    print('Последние 3 строки таблицы')
    display(dataset.tail(3))

    print()
    print('Информация о таблице')   
    print(dataset.info())

    print()
    print('Характеристики чисел в таблице')  
    display(dataset.describe())
      
    print()
    print('Переименование колонок')
def rename(dataset):
    dataset = dataset.rename(columns=lambda c: c.lower())
    return display(dataset.head(3))    

In [74]:
first_check(founders)
rename(founders)


Первые 3 строки таблицы


Name    Birthday Sex              email  \
0  Justin Graves  1953-03-06   F       it@gmail.com   
1   Tony Holland  1944-03-09   F  quality@gmail.com   
2    Angela Hall  1938-06-08   F   accept@gmail.com   

                                             Address            Company  \
0         34119 Garrett Spring\nWest Julie, AK 21587  Espinoza and Sons   
1  72475 Martin Village Apt. 623\nMorenobury, MA ...   Gilbert and Sons   
2          553 Harris Canyon\nShaunchester, SC 54399         Parker PLC   

                       Profession  Salary_USD$ foundation_date  city_id  \
0           Adult guidance worker         3904      2019-03-29   119164   
1  Furniture conservator/restorer         4678      2018-10-20   121306   
2             Visual merchandiser         3501      2017-05-16   112431   

    city_name  state_id state_code state_name                  Funs_club  
0   Jefferson      1401         MD   Maryland     Alternative Health NYC  
1    Marshall      1427         VA   Virginia  Alternative Energy Meetup  
2  Blissfield      1426         MI   Michigan          NYC Animal Rights


Последние 3 строки таблицы


Name    Birthday Sex              email  \
16387   Jason Carter  1967-11-04   N  culture@gmail.com   
16388   Noah Cochran  1995-10-11   M  benefit@gmail.com   
16389  Sherry Arroyo  1996-06-02   F   notice@gmail.com   

                                                 Address       Company  \
16387                   Unit 4187 Box 4723\nDPO AP 78372  Arnold-Brown   
16388  5195 French Motorway Apt. 500\nWest Jonathansh...     Potts Ltd   
16389              83811 Jared Neck\nLopezland, KY 44675    Barnes Inc   

                   Profession  Salary_USD$ foundation_date  city_id  \
16387     Art gallery manager         3189      2019-09-27   141252   
16388  Chief Strategy Officer         3717      2019-08-02   118219   
16389  Geologist, engineering         4214      2019-12-06   119210   

           city_name  state_id state_code  state_name Funs_club  
16387  Crystal Lakes      4851         OH        Ohio       NaN  
16388          Healy      1400         AK      Alaska       NaN  
16389   Jemez Pueblo      1423         NM  New Mexico       NaN


Информация о таблице
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16390 entries, 0 to 16389
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Name             16390 non-null  object
 1   Birthday         16390 non-null  object
 2   Sex              16390 non-null  object
 3   email            16390 non-null  object
 4   Address          16390 non-null  object
 5   Company          16390 non-null  object
 6   Profession       16390 non-null  object
 7   Salary_USD$      16390 non-null  int64 
 8   foundation_date  16390 non-null  object
 9   city_id          16390 non-null  int64 
 10  city_name        16390 non-null  object
 11  state_id         16390 non-null  int64 
 12  state_code       16390 non-null  object
 13  state_name       16390 non-null  object
 14  Funs_club        16328 non-null  object
dtypes: int64(3), object(12)
memory usage: 1.9+ MB
None

Характеристики чисел в таблице


Salary_USD$        city_id      state_id
count  16390.000000   16390.000000  16390.000000
mean    3244.707322  121551.227944   1568.693350
std     1008.582019    7159.149798    669.921004
min     1500.000000  110965.000000   1399.000000
25%     2369.000000  115988.250000   1418.000000
50%     3236.000000  120979.000000   1433.000000
75%     4122.750000  125852.000000   1451.000000
max     4999.000000  153289.000000   4851.000000


Переименование колонок


name    birthday sex              email  \
0  Justin Graves  1953-03-06   F       it@gmail.com   
1   Tony Holland  1944-03-09   F  quality@gmail.com   
2    Angela Hall  1938-06-08   F   accept@gmail.com   

                                             address            company  \
0         34119 Garrett Spring\nWest Julie, AK 21587  Espinoza and Sons   
1  72475 Martin Village Apt. 623\nMorenobury, MA ...   Gilbert and Sons   
2          553 Harris Canyon\nShaunchester, SC 54399         Parker PLC   

                       profession  salary_usd$ foundation_date  city_id  \
0           Adult guidance worker         3904      2019-03-29   119164   
1  Furniture conservator/restorer         4678      2018-10-20   121306   
2             Visual merchandiser         3501      2017-05-16   112431   

    city_name  state_id state_code state_name                  funs_club  
0   Jefferson      1401         MD   Maryland     Alternative Health NYC  
1    Marshall      1427         VA   Virginia  Alternative Energy Meetup  
2  Blissfield      1426         MI   Michigan          NYC Animal Rights

In [75]:
# переименуем название колонок в нижнем регистре
founders = founders.rename(columns=lambda c: c.lower())

следует сделать следующие преобразования в объекте founders:
- преобразовать строки Birthday и foundation_date в метки даты
- преобразовать числа  city_id  и  state_id  в строки
- преобразовать строки Sex в категории
- столбец Funs_club содержит значения NaN

In [76]:
founders['birthday'] = pd.to_datetime(founders['birthday'])
founders['foundation_date'] = pd.to_datetime(founders['foundation_date'])
founders['city_id'] = founders['city_id'].astype('str')
founders['state_id'] = founders['state_id'].astype('str')
founders['sex'] = founders['sex'].astype('category')
founders[:3]

name   birthday sex              email  \
0  Justin Graves 1953-03-06   F       it@gmail.com   
1   Tony Holland 1944-03-09   F  quality@gmail.com   
2    Angela Hall 1938-06-08   F   accept@gmail.com   

                                             address            company  \
0         34119 Garrett Spring\nWest Julie, AK 21587  Espinoza and Sons   
1  72475 Martin Village Apt. 623\nMorenobury, MA ...   Gilbert and Sons   
2          553 Harris Canyon\nShaunchester, SC 54399         Parker PLC   

                       profession  salary_usd$ foundation_date city_id  \
0           Adult guidance worker         3904      2019-03-29  119164   
1  Furniture conservator/restorer         4678      2018-10-20  121306   
2             Visual merchandiser         3501      2017-05-16  112431   

    city_name state_id state_code state_name                  funs_club  
0   Jefferson     1401         MD   Maryland     Alternative Health NYC  
1    Marshall     1427         VA   Virginia  Alternative Energy Meetup  
2  Blissfield     1426         MI   Michigan          NYC Animal Rights

In [77]:
founders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16390 entries, 0 to 16389
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             16390 non-null  object        
 1   birthday         16390 non-null  datetime64[ns]
 2   sex              16390 non-null  category      
 3   email            16390 non-null  object        
 4   address          16390 non-null  object        
 5   company          16390 non-null  object        
 6   profession       16390 non-null  object        
 7   salary_usd$      16390 non-null  int64         
 8   foundation_date  16390 non-null  datetime64[ns]
 9   city_id          16390 non-null  object        
 10  city_name        16390 non-null  object        
 11  state_id         16390 non-null  object        
 12  state_code       16390 non-null  object        
 13  state_name       16390 non-null  object        
 14  funs_club        16328 non-null  objec

In [82]:
founders.funs_club.value_counts(dropna=False)

NaN                                              62
Alternative Health NYC                            1
Offbeat Girlfriends                               1
New York Ping Pong Meetup - SPiN NY               1
Product Hacking                                   1
                                                 ..
Founders Space                                    1
NYC Synthmeet                                     1
The Italian Boot Camp: Language & Food Lovers     1
Wives of Virtue                                   1
FREE Arabic Chicago Evanston North Burbs          1
Name: funs_club, Length: 16329, dtype: int64

In [93]:
founders.describe()

salary_usd$
count  16328.000000
mean    3244.117528
std     1008.683939
min     1500.000000
25%     2368.000000
50%     3236.000000
75%     4122.000000
max     4999.000000

In [84]:
# В дальнейшем исследовании будут неинтересны строки с отсутствующими значениями funs_club, поэтому избавимся от них
founders = founders.dropna(subset = ["funs_club"])
founders

name   birthday sex                   email  \
0        Justin Graves 1953-03-06   F            it@gmail.com   
1         Tony Holland 1944-03-09   F       quality@gmail.com   
2          Angela Hall 1938-06-08   F        accept@gmail.com   
3      Brittany Larson 1951-09-19   N  professional@gmail.com   
4      Matthew Leblanc 1949-12-19   F         store@gmail.com   
...                ...        ...  ..                     ...   
16325     Karina Brown 1969-12-07   F       billion@gmail.com   
16326      Brian Meyer 1938-12-09   F         wrong@gmail.com   
16327    Victor Cooper 1963-06-27   F         argue@gmail.com   
16328        Cory Cain 2004-09-22   N        doctor@gmail.com   
16329    Walter Garcia 1966-09-07   N         truth@gmail.com   

                                                 address  \
0             34119 Garrett Spring\nWest Julie, AK 21587   
1      72475 Martin Village Apt. 623\nMorenobury, MA ...   
2              553 Harris Canyon\nShaunchester, SC 54399   
3                 2075 Wagner Fort\nAnitaville, ID 25417   
4                 444 Allen Ports\nEast Tanner, MD 02856   
...                                                  ...   
16325  333 Anthony Branch Suite 664\nLake Josephside,...   
16326   11866 Rodriguez Well\nNew Heatherburgh, LA 19080   
16327              7396 Sharon Tunnel\nShawton, TX 43883   
16328            449 Harris Station\nChadmouth, NJ 61397   
16329          916 Heidi Estates\nGregorymouth, VT 69272   

                          company                      profession  \
0               Espinoza and Sons           Adult guidance worker   
1                Gilbert and Sons  Furniture conservator/restorer   
2                      Parker PLC             Visual merchandiser   
3                   Gutierrez Ltd            Broadcast journalist   
4                     Owens-Jones   Restaurant manager, fast food   
...                           ...                             ...   
16325  Johnson, Torres and Acosta         Clinical cytogeneticist   
16326                 Woods-Olson     Nature conservation officer   
16327                  Arnold Inc             Trade mark attorney   
16328  Edwards, Allen and Bennett                         Barista   
16329              Walter-Webster             Early years teacher   

       salary_usd$ foundation_date city_id                city_name state_id  \
0             3904      2019-03-29  119164                Jefferson     1401   
1             4678      2018-10-20  121306                 Marshall     1427   
2             3501      2017-05-16  112431               Blissfield     1426   
3             2063      2020-01-14  125720  Sainte Genevieve County     1451   
4             3734      2019-12-20  116198              Essex Fells     1417   
...            ...             ...     ...                      ...      ...   
16325         4682      2020-03-25  114641  Cordry Sweetwater Lakes     1440   
16326         2828      2018-03-10  126231              Sharon Hill     1422   
16327         4430      2017-05-31  141760                     Troy     4851   
16328         3143      2018-05-17  141735                     Stow     4851   
16329         3215      2017-07-23  115415                   Dothan     1456   

      state_code    state_name  \
0             MD      Maryland   
1             VA      Virginia   
2             MI      Michigan   
3             MO      Missouri   
4             NJ    New Jersey   
...          ...           ...   
16325         IN       Indiana   
16326         PA  Pennsylvania   
16327         OH          Ohio   
16328         OH          Ohio   
16329         AL       Alabama   

                                               funs_club  
0                                 Alternative Health NYC  
1                              Alternative Energy Meetup  
2                                      NYC Animal Rights  
3                          The New York City Anime Group  
4                     

In [85]:
founders[['name', 'funs_club']].nunique()

name         14878
funs_club    16328
dtype: int64

В итоге имеем список из 16329 уникальных названий клубов и 14879 уникальных владельцев/основателей (1450 человек основали ведут несколько клубов)

In [91]:
# при загрузке сразу переведем некоторые данные в формат datetime
members = pd.read_csv('members.csv', parse_dates = (['Birthday', 'Entry_date']))

In [92]:
first_check(members)
rename(members)


Первые 3 строки таблицы


Name   Birthday Sex            email  \
0    Daniel Bryan 1950-09-08   N   trip@gmail.com   
1  Crystal Galvan 1956-09-29   M    now@gmail.com   
2  Jose Rodriguez 1982-02-14   N  style@gmail.com   

                                             Address                 Company  \
0     71126 Hammond Estate\nNew Juliehaven, MN 79973         Ramirez-Anthony   
1                   PSC 9268, Box 5767\nAPO AP 37347           Curtis-Taylor   
2  5659 Michael Centers Apt. 094\nSouth Nicolesta...  Smith, Tucker and Ryan   

            Profession  Salary_USD$ Entry_date  group_id  city_id   city_name  \
0      Mining engineer         4077 2017-09-21  25667956   126108  Sebastopol   
1  Exhibition designer         2610 2018-06-07  15795992   141718    Somerset   
2        Archaeologist         4571 2017-08-12   1753199   123760      Palmer   

   state_id state_code  state_name  
0      1416         CA  California  
1      4851         OH        Ohio  
2      1400         AK      Alaska


Последние 3 строки таблицы


Name   Birthday Sex              email  \
144830      Zachary Braun 1944-09-24   F     wait@gmail.com   
144831  Tracey Benitez MD 1940-07-19   M     name@gmail.com   
144832        Amy Merritt 1996-04-01   F  perhaps@gmail.com   

                                                  Address       Company  \
144830  2708 Barr Ford Suite 255\nPort Staceyshire, SC...  Mitchell Ltd   
144831           0496 John Meadows\nDouglasview, MN 08568   Jones-Watts   
144832  98796 Gentry Village Suite 224\nDonnaport, FL ...   Eaton Group   

                Profession  Salary_USD$ Entry_date  group_id  city_id  \
144830  Jewellery designer         3670 2017-07-25  12412382   128181   
144831      Engineer, land         4581 2018-05-28  10730622   112671   
144832    Doctor, hospital         3323 2017-10-19   1766833   124402   

        city_name  state_id state_code     state_name  
144830  Verdigris      1421         OK       Oklahoma  
144831    Bradley      1429         WV  West Virginia  
144832   Plymouth      1433         MA  Massachusetts


Информация о таблице
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144833 entries, 0 to 144832
Data columns (total 15 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Name         144833 non-null  object        
 1   Birthday     144833 non-null  datetime64[ns]
 2   Sex          144833 non-null  object        
 3   email        144833 non-null  object        
 4   Address      144833 non-null  object        
 5   Company      144833 non-null  object        
 6   Profession   144833 non-null  object        
 7   Salary_USD$  144833 non-null  int64         
 8   Entry_date   144833 non-null  datetime64[ns]
 9   group_id     144833 non-null  int64         
 10  city_id      144833 non-null  int64         
 11  city_name    144833 non-null  object        
 12  state_id     144833 non-null  int64         
 13  state_code   144833 non-null  object        
 14  state_name   144833 non-null  object        
dtypes: datetime6

Salary_USD$      group_id        city_id       state_id
count  144833.000000  1.448330e+05  144833.000000  144833.000000
mean     3249.352627  1.554929e+07  121558.013588    1570.038182
std      1009.090695  8.812839e+06    7163.339427     673.052612
min      1500.000000  6.388000e+03  110965.000000    1399.000000
25%      2377.000000  6.695442e+06  115989.000000    1418.000000
50%      3250.000000  1.897946e+07  121022.000000    1433.000000
75%      4122.000000  2.259329e+07  125874.000000    1451.000000
max      4999.000000  2.637847e+07  153289.000000    4851.000000


Переименование колонок


name   birthday sex            email  \
0    Daniel Bryan 1950-09-08   N   trip@gmail.com   
1  Crystal Galvan 1956-09-29   M    now@gmail.com   
2  Jose Rodriguez 1982-02-14   N  style@gmail.com   

                                             address                 company  \
0     71126 Hammond Estate\nNew Juliehaven, MN 79973         Ramirez-Anthony   
1                   PSC 9268, Box 5767\nAPO AP 37347           Curtis-Taylor   
2  5659 Michael Centers Apt. 094\nSouth Nicolesta...  Smith, Tucker and Ryan   

            profession  salary_usd$ entry_date  group_id  city_id   city_name  \
0      Mining engineer         4077 2017-09-21  25667956   126108  Sebastopol   
1  Exhibition designer         2610 2018-06-07  15795992   141718    Somerset   
2        Archaeologist         4571 2017-08-12   1753199   123760      Palmer   

   state_id state_code  state_name  
0      1416         CA  California  
1      4851         OH        Ohio  
2      1400         AK      Alaska

In [95]:
# переименуем название колонок в нижнем регистре
members = members.rename(columns=lambda c: c.lower())
members.sample(3)

name   birthday sex              email  \
63371    Richard Andersen 1990-02-14   M  history@gmail.com   
139062  Elizabeth Pacheco 1954-10-10   M    mouth@gmail.com   
132033         Drew Mckay 1962-11-23   F     fire@gmail.com   

                                                address            company  \
63371   8326 Zoe Falls Apt. 055\nBurnettshire, CT 35807       Smith-Vaughn   
139062          5271 James Estate\nYoungville, ME 36350       Hoover Group   
132033   3773 Miller Hollow\nWest Jessicafort, ID 09455  Oconnell-Trujillo   

                             profession  salary_usd$ entry_date  group_id  \
63371                       Pathologist         1803 2019-03-04  17774602   
139062            Accommodation manager         3404 2018-12-03  20481783   
132033  Psychotherapist, dance movement         4714 2017-09-23   1940661   

        city_id     city_name  state_id state_code  state_name  
63371    125544  Russellville      1419         KY    Kentucky  
139062   121501       McGehee      1444         AR    Arkansas  
132033   113800       Chatham      1417         NJ  New Jersey

In [98]:
members['city_id'] = members['city_id'].astype('str')
members['state_id'] = members['state_id'].astype('str')
members['group_id'] = members['group_id'].astype('str')
founders['sex'] = founders['sex'].astype('category')
founders[:3]

name   birthday sex              email  \
0  Justin Graves 1953-03-06   F       it@gmail.com   
1   Tony Holland 1944-03-09   F  quality@gmail.com   
2    Angela Hall 1938-06-08   F   accept@gmail.com   

                                             address            company  \
0         34119 Garrett Spring\nWest Julie, AK 21587  Espinoza and Sons   
1  72475 Martin Village Apt. 623\nMorenobury, MA ...   Gilbert and Sons   
2          553 Harris Canyon\nShaunchester, SC 54399         Parker PLC   

                       profession  salary_usd$ foundation_date city_id  \
0           Adult guidance worker         3904      2019-03-29  119164   
1  Furniture conservator/restorer         4678      2018-10-20  121306   
2             Visual merchandiser         3501      2017-05-16  112431   

    city_name state_id state_code state_name                  funs_club  
0   Jefferson     1401         MD   Maryland     Alternative Health NYC  
1    Marshall     1427         VA   Virginia  Alternative Energy Meetup  
2  Blissfield     1426         MI   Michigan          NYC Animal Rights

In [99]:
members.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144833 entries, 0 to 144832
Data columns (total 15 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   name         144833 non-null  object        
 1   birthday     144833 non-null  datetime64[ns]
 2   sex          144833 non-null  object        
 3   email        144833 non-null  object        
 4   address      144833 non-null  object        
 5   company      144833 non-null  object        
 6   profession   144833 non-null  object        
 7   salary_usd$  144833 non-null  int64         
 8   entry_date   144833 non-null  datetime64[ns]
 9   group_id     144833 non-null  object        
 10  city_id      144833 non-null  object        
 11  city_name    144833 non-null  object        
 12  state_id     144833 non-null  object        
 13  state_code   144833 non-null  object        
 14  state_name   144833 non-null  object        
dtypes: datetime64[ns](2), int64(1), ob

In [100]:
members.describe()

salary_usd$
count  144833.000000
mean     3249.352627
std      1009.090695
min      1500.000000
25%      2377.000000
50%      3250.000000
75%      4122.000000
max      4999.000000

In [103]:
# загрузим оставшиеся датасеты
hobby_list = pd.read_csv('hobby_list.csv')
categories = pd.read_csv('categories.csv')

In [104]:
first_check(hobby_list)


Первые 3 строки таблицы


group_id                       name  category_id  monthly_fee_USD$
0      6388     Alternative Health NYC           14              13.0
1      6510  Alternative Energy Meetup            4               7.0
2      8458          NYC Animal Rights           26              15.0


Последние 3 строки таблицы


group_id                                      name  category_id  \
16326  26378067               Streeterville Running Group            9   
16327  26378128                            Just Dance NYC           23   
16328  26378470  FREE Arabic Chicago Evanston North Burbs           31   

       monthly_fee_USD$  
16326               NaN  
16327               NaN  
16328               NaN


Информация о таблице
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16329 entries, 0 to 16328
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   group_id          16329 non-null  int64  
 1   name              16328 non-null  object 
 2   category_id       16329 non-null  int64  
 3   monthly_fee_USD$  16315 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 510.4+ KB
None

Характеристики чисел в таблице


group_id   category_id  monthly_fee_USD$
count  1.632900e+04  16329.000000      16315.000000
mean   1.553368e+07     19.603956         12.012994
std    8.813184e+06     12.373108          7.226782
min    6.388000e+03      1.000000          0.000000
25%    6.647792e+06      8.000000          6.000000
50%    1.897088e+07     21.000000         12.000000
75%    2.259329e+07     33.000000         18.000000
max    2.637847e+07     36.000000         24.000000


Переименование колонок


In [106]:
categories

category_id               category_name
0             1              Arts & Culture
1             3          Cars & Motorcycles
2             4     Community & Environment
3             5                     Dancing
4             6        Education & Learning
5             9                     Fitness
6            10                Food & Drink
7            11                       Games
8            14          Health & Wellbeing
9            15            Hobbies & Crafts
10           16  Language & Ethnic Identity
11           17                   Lifestyle
12           18                  Book Clubs
13           20               Movies & Film
14           21                       Music
15           22      New Age & Spirituality
16           23        Outdoors & Adventure
17           24                  Paranormal
18           25            Parents & Family
19           27                 Photography
20           28          Religion & Beliefs
21           29            Sci-Fi & Fantasy
22           30                     Singles
23           32         Sports & Recreation
24           33                     Support
25           36                     Writing
26           37               Miscellaneous